### PR TITLE
get first stack trace info.

### PR DESCRIPTION
**Pull request by wangsai-silence**
For this issue https://github.com/pkg/errors/issues/173

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -280,3 +280,34 @@ func Cause(err error) error {
...
+	}
+
+	var stackErr error
+
+	for {
+		if _, ok := err.(stackTracer); ok {
+			stackErr = err
+		}
+
+		if causer, ok := err.(causer); ok {
```
Comment (puellanivis) [inline errors.go]: Prefer:
```
		cause, ok := err.(causer)
		if !ok {
			break
		}
		err = cause.Cause()
```
The same as `Cause` in this package.

File: errors.go [hunk 2]
```diff
@@ -280,3 +280,34 @@ func Cause(err error) error {
...
+func Stack(err error) StackTrace {
+	type causer interface {
+		Cause() error
+	}
+
+	type stackTracer interface {
+		StackTrace() StackTrace
+	}
+
+	var stackErr error
```
Comment (puellanivis) [inline errors.go]: We could avoid having to re-type-assert the stackTracer if we recognize that `stackErr` is always going to be a `stackTracer` and enforce that through typing:
```
	var top stackTracker
	…
		if stack, ok := err.(stackTracer); ok {
			top = stack
		}
	…
	if top == nil {
		return nil
	}

	return top.StackTrace()
```

File: errors_test.go [hunk 3]
```diff
@@ -106,6 +106,54 @@ func TestCause(t *testing.T) {
...
+	}
+
+	for i, tt := range tests {
+		got := Stack(tt.err)
+		if got == nil || tt.want == nil {
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("test %d: got %#v, want %#v", i+1, got, tt.want)
+			}
+
+			return
```
Comment (puellanivis) [inline errors_test.go]:
```suggestion
			continue
```

Otherwise, we break out of the test after the first test, because definitively `tt.want == nil`.

Also, as written, by contract of the code design, we should expect all the other table tests to fail after this change.

Comment (wangsai-silence) [reply errors_test.go]: You are right. I will rewrite it.

Comment (wangsai-silence) [reply errors_test.go]: Thank you for your comments

File: errors.go [hunk 4]
```diff
@@ -280,3 +280,36 @@ func Cause(err error) error {
 	}
 	return err
 }
+
+// Stack returns the first stack trace of the errors, if possible.
```
[the remark this answers is not in the document]
Comment (wangsai-silence) [reply errors.go]: Thank you so much for your patience.  Your advice has been invaluable, and I have followed the suggest expect the last one. I don't know how to deal with this documentation. And I just wrote your words on it. So could you help to make it more clear?

File: errors.go [hunk 5]
```diff
@@ -280,3 +280,40 @@ func Cause(err error) error {
 	}
 	return err
 }
+
+// Stack returns the earliest/deepest StackTrace attached to any of
+// the errors in the chain of Causes. If the error does not implement
+// Cause, the original error will be returned. If the error is nil,
+// nil will be returned without further investigation.
+
```
Comment (puellanivis) [inline errors.go]: The docstrings should not have a space between them an the function. I don’t know for sure this would break godoc, but convention suggest we strictly enforce it.

Comment (wangsai-silence) [reply errors.go]: OK. I'd remove it. Thank you.

File: errors.go [hunk 6]
```diff
@@ -280,3 +280,39 @@ func Cause(err error) error {
 	}
 	return err
 }
+
+// Stack returns the earliest/deepest StackTrace attached to any of
+// the errors in the chain of Causes. If the error does not implement
+// Cause, the original error will be returned. If the error is nil,
```
Comment (flimzy) [inline errors.go]: I think you mean "the stacktrace from the original error will be returned"?

File: errors.go [hunk 7]
```diff
@@ -280,3 +280,39 @@ func Cause(err error) error {
...
+// nil will be returned without further investigation.
+func Stack(err error) StackTrace {
+	type causer interface {
+		Cause() error
+	}
+
+	type stackTracer interface {
+		StackTrace() StackTrace
+	}
```
Comment (flimzy) [inline errors.go]: You shouldn't need to define these types here--these are already global types.

File: errors.go [hunk 8]
```diff
@@ -280,3 +280,39 @@ func Cause(err error) error {
...
+		}
+
+		err = causer.Cause()
+	}
+
+	if topStackInfo != nil {
+		return topStackInfo
+	}
+
+	return nil
```
Comment (flimzy) [inline errors.go]: You can collapse these 4 lines to simply:

    return topStackInfo

**Comment by hanzei** [comment]
@wangsai-silence Could you please take a look at the comments above?

File: errors.go (hunk 7 above)
Comment (wangsai-silence) [reply errors.go]: Sorry, but I didn't find any public define of these types. Only private ones.

**Comment by wangsai-silence** [comment]
> @wangsai-silence Could you please take a look at the comments above?

OK. I've done. Thank you for remind me about that.

File: errors.go (hunk 7 above)
Comment (flimzy) [reply errors.go]: This one is also private. Do you need a public one?

Comment (wangsai-silence) [reply errors.go]: I'm not sure about that. I don't know why the previous ones are all private.

Comment (puellanivis) [reply errors.go]: We want to keep the scope of the interfaces to the smallest set available. Firstly, because defining these interfaces scoped to the function that uses them makes it more clear what the interface is (it’s defined right above you).  Secondly, it also prevents us from exporting interface types unnecessarily (one cannot export something that is not even there) Finally, there’s considerations that we should be avoiding using interfaces as much as possible. Giving some overhead required to actually make use of them is a good way to deter rampant interface usage.

Basically, we should generally be avoiding defining package-global interfaces as much as possible.

**Comment by mightyguava** [comment]
Oh I was just looking to implement this. @wangsai-silence @puellanivis what is left to be done for this PR?

**Comment by wangsai-silence** [comment]
> Oh I was just looking for something like this. @wangsai-silence @puellanivis what is left to be done for this PR?

Nobody answered. I have used it in my project.

**Comment by puellanivis** [comment]
🤷 I don’t have collaborator rights to the repo, so I cannot approve and merge. Also, I was going to leaving the approval/denial of the mere design up to other parties anyways.

The branch is however now out-of-date with the base branch, and likely would end up with a merge conflict or something.

Fortunately, the code works both inside of and outside of this package. So there is not much stopping anyone from replicating it. But, with proper `pkg/error` use sanitation, you can avoid the need to use it drastically.

**Comment by flimzy** [comment]
FWIW, I [proposed something like this](https://github.com/pkg/errors/issues/94) back in 2016. While it wasn't outright rejected, I closed the issue because I agreed with Dave Cheney's cautious approach.

**Comment by mightyguava** [comment]
@flimzy ah, asking each package to implement their own version makes sense with that link. I don't understand what `errors.As` has to do with this though. I want to print a stacktrace for debugging.

**Comment by flimzy** [comment]
> @flimzy ah, asking each package to implement their own version makes sense with that link. I don't understand what `errors.As` has to do with this though. I want to print a stacktrace for debugging.

Two things:

1. I want a version that honors Go 1.13's `wrapper` interface. `errors.As` is the easiest way to address that.
2. `errors.As` makes it trivial to implement this functionality, making the need for supporting it in a library like this less urgent, IMO. Example (untested "whiteboard" code):

```golang
stackErr := new(interface{ StackTrace() errors.StackTrace })
for errors.As(err, stackErr) {
    err = stackErr
}
if stack != nil {
    stack := stackErr.StackTrace()
}
```

**Comment by mightyguava** [comment]
`errors.As` doesn't unwrap unless necessary, so I think you have an infinite loop there.

**Comment by flimzy** [comment]
> `errors.As` doesn't unwrap unless necessary, so I think you have an infinite loop there.

Possible. As I said, the code is back-of-napkin quality, untested. I don't think it affects the substance of what I was saying.

**Comment by puellanivis** [comment]
You are correct, `errors.As` does not unwrap if the given `error` is already assignable to the second argument. That’s easy enough to fix with an `errors.Unwrap` call.

But I would additionally recommend against using `new()` for an interface. Because an interface is already reference semantics, it is almost universally an error to make a direct datatype that is a pointer to an interface. The only time a pointer to interface is not an error, is when it is a derived type, for instance when you have a subfunction that would like to assign into an interface value, rather than just use it, where it should be derefed at the subfunction call. So, these changes get us:
```
func Stack(err error) StackTrace {
	var stackErr interface{ StackTrace() errors.StackTrace }
	for errors.As(err, &stackErr) {
		err = errors.Unwrap(stackErr) // err will be assignable, and infinite loop otherwise.
	}
	if stackErr != nil {
		return stackErr.StackTrace()
	}
	return nil
}
```

**Comment by flimzy** [comment]
> But I would additionally recommend against using `new()` for an interface. <snip>

All good advice. Please don't get hung up on the minutiae of a throw-away coding example.

**Comment by puellanivis** [comment]
I don’t think we derailed things. We ended up with a corrected version, that people can use. It’s the good-ol’ “say a wrong answer, and you will immediately get 50 other people to do the work of finding the correct answer for you.”
